kevlar/simplex.py

```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python
+#
+# -----------------------------------------------------------------------------
+# Copyright (c) 2017 The Regents of the University of California
+#
+# This file is part of kevlar (http://github.com/dib-lab/kevlar) and is
+# licensed under the MIT license: see LICENSE.
+# -----------------------------------------------------------------------------
+
+import sys
+import kevlar
+from kevlar.novel import novel, load_samples
+from kevlar.filter import filter as kfilter
+from kevlar.filter import load_mask
+from kevlar.partition import partition
+from kevlar.alac import alac
+
+
+def simplex(case, casecounts, controlcounts, refrfile, ksize=31, ctrlmax=0,
+            casemin=5, mask=None, filtermem=1e6,
+            filterfpr=0.001, partminabund=2, partmaxabund=200, dedup=True,
+            delta=50, match=1, mismatch=2, gapopen=5, gapextend=0,
+            greedy=False, logstream=sys.stderr):
+    """
+    Execute the simplex germline variant discovery workflow.
+
+    FIXME
+    """
+    discoverer = novel(
+        case, [casecounts], controlcounts, ksize=ksize, casemin=casemin,
+        ctrlmax=ctrlmax, logstream=logstream
+    )
+    filterer = kfilter(
+        discoverer, mask=mask, minabund=casemin, ksize=ksize,
+        memory=filtermem, maxfpr=filterfpr, logstream=logstream
+    )
+    partitioner = partition(
+        filterer, dedup=dedup, minabund=partminabund, maxabund=partmaxabund,
+        logstream=logstream
+    )
+    for cc in partitioner:
+        caller = alac(
+            cc, refrfile, ksize=ksize, delta=delta, match=match,
+            mismatch=mismatch, gapopen=gapopen, gapextend=gapextend,
+            greedy=greedy, logstream=logstream
+        )
+        for variant in caller:
+            yield variant
+
+
+def main(args):
+    cases = load_samples(
+        args.case_counts, [args.case], ksize=args.ksize,
+        memory=args.novel_memory, maxfpr=args.novel_fpr,
+        numthreads=args.threads, logstream=args.logfile
+    )
+    controls = load_samples(
+        args.control_counts, args.control, ksize=args.ksize,
+        memory=args.novel_memory, maxfpr=args.novel_fpr,
+        numthreads=args.threads, logstream=args.logfile
+    )
+    mask = load_mask(
+        args.mask_files, args.ksize, args.mask_memory, maxfpr=args.filter_fpr,
+        logstream=args.logfile
+    )
+
+    outstream = kevlar.open(args.out, 'w')
+    caserecords = kevlar.multi_file_iter_screed(args.case)
+    workflow = simplex(
+        caserecords, cases[0], controls, args.refr, ksize=args.ksize,
```

Comment:
Using `cases[0]` instead of `cases` is compensated by using `[casecounts]` when calling `kevlar.novel`. This is adding unnecessary complexity. The array (length 1) of counttables should simply be passed to the `kevlar.novel` function.

Reply:
I take it back. Users of the Python API are much more likely to be calling the `simplex` workflow than any of its smaller subcomponents, so I'm optimizing for clarity of that interface.